### PR TITLE
feat: profiles caching

### PR DESCRIPTION
**Pull request by arturtkachenko**

### CODE REVIEW

File: UP Agent/content/wallet/transfer/AddressDropdown/AddressDropdownRow.swift [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+//  UP Agent
+//
+//  Created by Artur Tkachenko on 20.12.21.
+//
+
+import SwiftUI
+
+struct AddressDropdownRow: View {
+    
+    var address: String
```
Comment (JeneaVranceanu) [inline UP Agent/content/wallet/transfer/AddressDropdown/AddressDropdownRow.swift:12]: Consider adding `IdenticonData` to display it in the dropdown.
Again, it is just a suggestion but will be a good visual cue to help recognize the profile you wish to select.

File: UP Agent/content/wallet/transfer/AddressDropdown/AddressDropdownView.swift [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+    
+    var addresses: [String]
+    var onAddressSelected: ((_ address: String) -> Void)?
+    
+    var body: some View {
+        let minHeight = CGFloat((addresses.count < 5) ? (addresses.count * 50) : 250)
+        ScrollView {
+            VStack(alignment: .leading, spacing: .zero) {
+                ForEach(self.addresses, id: \.self) { address in
+                    DropdownRow(address: address, onAddressSelected: self.onAddressSelected)
```
[the remark this answers is not in the document]
Comment (JeneaVranceanu) [reply UP Agent/content/wallet/transfer/AddressDropdown/AddressDropdownView.swift]: And you can drop `self` from `self.onAddressSelected`. That I didn't notice right away )

File: UP Agent/content/wallet/transfer/TransferView.swift [hunk 3]
```diff
@@ -198,23 +200,38 @@ struct TransferView: View {
...
+                        shouldShowProfileSuggestions = isEditing
+                    },
+                    onCommit: {
                         viewModel.fetchProfile(address)
+                    })
+                    .keyboardType(UIKeyboardType.webSearch)
+                    .autocapitalization(.none)
+                    .frame(maxWidth: .infinity)
+                        
+                        createTextButton(label: "Search") {
```
Comment (JeneaVranceanu) [inline UP Agent/content/wallet/transfer/TransferView.swift:226]: A little bit of code formatting would be nice here (indents are off a little bit).

File: UP Agent/content/wallet/transfer/TransferView.swift [hunk 4]
```diff
@@ -198,23 +200,38 @@ struct TransferView: View {
...
+                        }
+                        .frame(width: 60)
+                        .padding(.leading, 8)
                     }
-                    .frame(width: 60)
-                    .padding(.leading, 8)
+                
+                if shouldShowProfileSuggestions && !viewModel.searchedProfiles().isEmpty {
+                    AddressDropdownView(addresses: viewModel.searchedProfiles(by: address),
+                                 onAddressSelected: { address in
```
Comment (JeneaVranceanu) [inline UP Agent/content/wallet/transfer/TransferView.swift:236]: Same here with indents.

File: UP Agent/content/wallet/transfer/viewmodel/TransferViewModel.swift [hunk 5]
```diff
@@ -56,6 +59,14 @@ class TransferViewModel: ObservableObject {
         fetchBalance()
     }
     
+    func searchedProfiles(by address: String = "") -> [String] {
+        guard !address.isEmpty else {
+            return userProxy.searchedProfiles
+        }
+        
+        return userProxy.searchedProfiles.filter { $0.contains(address) }
```
Comment (JeneaVranceanu) [inline UP Agent/content/wallet/transfer/viewmodel/TransferViewModel.swift:67]: Consider making this comparison case insensitive. It is unlikely users will remember a checksummed version of an ERC725 address :)

But the ones we present and fill inputs with after a selection has happened should always be checksummed.

File: UP Agent/data/model/UserProfilesProxy.swift [hunk 6]
```diff
@@ -24,22 +24,39 @@ class UserProfilesProxy: ObservableObject {
...
     var address: String? {
         currentProfile?.address
     }
+    
     var lsp3Profile: LSP3Profile? {
         currentProfile?.lsp3Profile
     }
+    
+    var searchedLSP3Profiles: [LSP3Profile]? {
+        []
```
Comment (JeneaVranceanu) [inline UP Agent/data/model/UserProfilesProxy.swift]: With activities implemented we now have RealmSwift DB as a dependency. If you want to you can create a separate database file where all LSP3Profiles we fetch will be stored. 

What do you think about it?

Comment (arturtkachenko) [reply UP Agent/data/model/UserProfilesProxy.swift]: Oops, looks like I forgot to delete this property, I don't use LSP3Profile later.

File: UP Agent/content/wallet/transfer/viewmodel/TransferViewModel.swift (hunk 5 above)
Comment (arturtkachenko) [reply UP Agent/content/wallet/transfer/viewmodel/TransferViewModel.swift:67]: Sure, make sense.

**Pull request merged**